### PR TITLE
[Feature:Forum] Forum Autosave

### LINKED ISSUE #5449: save contents of forum post to local storage
**What problem are you trying to solve with Submitty**
If the user has written a lengthy forum post in the input form box, but clicks away before submitting to do another task then returns to the page to submit they may lose that writing.

**Describe the way you'd like to solve this problem**
Similar to the change we made for notebook gradeable input boxes (issue #5238 & PR #5323), let's save the contents of the box to browser local storage.  So on page refresh they don't lose work.

### PR DESCRIPTION
**Pull request by hdrodz97**
### Please check if the PR fulfills these requirements:

* [X] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [X] Tests for the changes have been added/updated (if possible)
* [X] Documentation has been updated/added if relevant

### What is the current behavior?

Closes #5449

If a user typed a lengthy post in the forum (either as a response to some post or creating a new thread) and accidentally leaves the page, all of their progress is lost.

### What is the new behavior?

Similar to notebook gradeables (see #5323), we implement autosaving to localStorage. Each reply box has autosave data associated with it (which will only be saved if the box is not empty). When a reply is posted, then that reply box's autosave data will be cleared on a successful submission.

Additionally, this PR implements autosave functionality for the "Create a thread" feature, saving the thread title, categories, text, status, and anonymity.

### Other information?

Tested on Firefox on Linux and on Safari on macOS.

### CODE REVIEW

File: site/app/templates/forum/createThread.twig [hunk 1]
```diff
@@ -42,9 +42,67 @@
...
+                }
+                $(e).trigger("eventChangeCatClass");
+            });
+        }
+    }
+
+    function clearCreateThreadAutosave() {
+        localStorage.removeItem(CREATE_THREAD_AUTOSAVE_KEY);
+    }
+
```
Comment (andrewaikens87) [inline site/app/templates/forum/createThread.twig:91]: Is there a reason that this isn't in forum.js?

Comment (hdrodz97) [reply site/app/templates/forum/createThread.twig]: Not really, moved

File: site/public/js/forum.js [hunk 2]
```diff
@@ -1420,3 +1426,95 @@ function updateSelectedThreadContent(selected_thread_first_post_id){
         }
     });
 }
+
+const FORUM_AUTOSAVE_KEY = `${window.location.pathname}-forum-autosave`;
+
+function autosaveKeyFor(textarea) {
+    return `${window.location.pathname}-${textarea.id}-forum-autosave`;
```
Comment (mukul-kmr-jha) [inline site/public/js/forum.js:1433]: So, for the general textarea... one shown in the bottom I think we should not be using `textarea.id` and instead, use something general word or something?

File: site/public/js/forum.js [hunk 3]
```diff
@@ -1420,3 +1426,95 @@ function updateSelectedThreadContent(selected_thread_first_post_id){
...
+function restoreCreateThreadFromLocal() {
+    if (autosaveEnabled) {
+        const json = localStorage.getItem(CREATE_THREAD_AUTOSAVE_KEY);
+        if (!json) {
+            return;
+        }
+
+        const { title, isAnonymous, categories, status } = JSON.parse(json);
+        $("#title").val(title);
+        $("#input.thread-anon-checkbox").prop("checked", isAnonymous);
```
Comment (mukul-kmr-jha) [inline site/public/js/forum.js:1498]:
```suggestion
        $("input.thread-anon-checkbox").prop("checked", isAnonymous);
```

File: site/app/templates/forum/createThread.twig [hunk 4]
```diff
@@ -44,7 +44,18 @@
...
+            cancelDeferredSave(CREATE_THREAD_DEFER_KEY);
+            clearCreateThreadAutosave();
+        });
         $("#thread_form").submit(createThread);
         $("form").areYouSure();
+
+        cleanupAutosaveHistory('-create-autosave');
+        restoreCreateThreadFromLocal();
+        $("#title").on("input", () => deferredSave(CREATE_THREAD_DEFER_KEY, saveCreateThreadToLocal, 1));
+        $("#input.thread-anon-checkbox").change(saveCreateThreadToLocal);
```
Comment (mukul-kmr-jha) [inline site/app/templates/forum/createThread.twig:57]:
```suggestion
        $("input.thread-anon-checkbox").change(saveCreateThreadToLocal);
```

**Comment by bmcutler** [comment]
@hdrodz97 
Please fix merge conflicts

File: site/public/js/forum.js (hunk 2 above)
Comment (hdrodz97) [reply site/public/js/forum.js:1932]: Yup, you're right, I'll fix this

File: site/app/templates/forum/createThread.twig [hunk 5]
```diff
@@ -44,8 +44,20 @@
...
+        });
         $("#thread_form").submit(createThread);
         $("form").areYouSure();
+        
+        cleanupAutosaveHistory('-create-autosave');
+        restoreCreateThreadFromLocal();
+        $("#title").on("input", () => deferredSave(CREATE_THREAD_DEFER_KEY, saveCreateThreadToLocal, 1));
+        $("input.thread-anon-checkbox").change(saveCreateThreadToLocal);
+        $("div.cat-buttons").click(saveCreateThreadToLocal);
+        $("#thread_status").change(saveCreateThreadToLocal);
```
Comment (mukul-kmr-jha) [inline site/app/templates/forum/createThread.twig:59]: Add listener for other fields as well.

**Pull request merged**